compiler/rustc_typeck/src/check/generator_interior.rs

```diff
@@ -208,66 +210,81 @@ pub fn resolve_interior<'a, 'tcx>(
-    // if a Sync generator contains an &'α T, we need to check whether &'α T: Sync),
-    // so knowledge of the exact relationships between them isn't particularly important.
 
-    debug!("types in generator {:?}, span = {:?}", types, body.value.span);
+    debug!("types in generator {:?}, span = {:?}", causes, body.value.span);
 
+    // Mapping from region -> anonymized bound region
     let mut counter = 0;
+    let mut region_to_anon = FxHashMap::default();
+    let mut anonymize_regions = |region, current_depth| {
+        let br = region_to_anon.entry(region).or_insert_with(|| {
+            let br =
+                ty::BoundRegion { var: ty::BoundVar::from_u32(counter), kind: ty::BrAnon(counter) };
+            counter += 1;
+            br
+        });
+        fcx.tcx.mk_region(ty::ReLateBound(current_depth, *br))
+    };
+
+    // Deduplicate the generator interior types, since we don't need several of the same type
     let mut captured_tys = FxHashSet::default();
-    let type_causes: Vec<_> = types
-        .into_iter()
-        .filter_map(|mut cause| {
-            // Erase regions and canonicalize late-bound regions to deduplicate as many types as we
-            // can.
-            let erased = fcx.tcx.erase_regions(cause.ty);
-            if captured_tys.insert(erased) {
-                // Replace all regions inside the generator interior with late bound regions.
-                // Note that each region slot in the types gets a new fresh late bound region,
-                // which means that none of the regions inside relate to any other, even if
-                // typeck had previously found constraints that would cause them to be related.
-                let folded = fcx.tcx.fold_regions(erased, &mut false, |_, current_depth| {
-                    let br = ty::BoundRegion {
-                        var: ty::BoundVar::from_u32(counter),
-                        kind: ty::BrAnon(counter),
-                    };
-                    let r = fcx.tcx.mk_region(ty::ReLateBound(current_depth, br));
-                    counter += 1;
-                    r
-                });
+    causes.retain(|cause| captured_tys.insert(cause.ty));
 
-                cause.ty = folded;
-                Some(cause)
-            } else {
-                None
-            }
+    // Extract type components to build the witness type, anonymizing regions
+    let anon_tys = fcx.tcx.mk_type_list(
+        causes
+            .iter()
+            .map(|cause| fcx.tcx.fold_regions(cause.ty, &mut false, &mut anonymize_regions)),
+    );
+
+    // Extract the projection types that we know hold in the interior types, so we can reapply them
+    // later. This step is important because when we erase regions, we might no longer be able to
+    // apply some projections (since they depend on region knowledge we erased). But since we know
+    // they hold now, they should hold forever, so we'll add these projections to the param-env later.
```

Comment:
It's not clear to me that this is true. The region relationships aren't really checked now, they are checked later, in MIR borrow check. I have to ponder whether I can make a failing test case, though, it may be that we wind up enforcing the resulting constraints in MIR borrow check as well. If not, I think we could fix it by "reproving" these predicates.

Reply:
> it may be that we wind up enforcing the resulting constraints in MIR borrow check as well

Yeah that's what I thought would happen here. What I thought was: when we normalize the type, it registers obligations into the fcx with the local (non-erased) region variables. If we had a bad projection type, then we would have a borrowck error regardless, so it's fine to treat this as "solved" when we check the generator interior later on.